molviewspec/molviewspec/nodes.py

```diff
@@ -39,17 +39,30 @@
 ]
 
 
+CustomT = Optional[Mapping[str, Any]]
+RefT = Optional[str]
+
+
 class Node(BaseModel):
     """
     Base impl of all state tree nodes.
     """
 
     kind: KindT = Field(description="The type of this node.")
     params: Optional[Mapping[str, Any]] = Field(description="Optional params that are needed for this node.")
-    additional_properties: Optional[Mapping[str, Any]] = Field(
-        description="Optional free-style dict with custom, non-schema props."
-    )
     children: Optional[list[Node]] = Field(description="Optional collection of nested child nodes.")
+    custom: CustomT = Field(description="Custom data to store attached to this node.")
+    ref: RefT = Field(description="Optional reference that can be used to access this node.")
+
+    def __init__(self, **data):
+        # extract `custom` value from `params`
+        params = data.get("params", {})
+        if "custom" in params:
+            data["custom"] = params.pop("custom")
+        if "ref" in params:
+            data["ref"] = params.pop("ref")
```

Comment:
@JonStargaryen  @midlik  Added support for `refs` as well which will be required to make the primitives works when showing lines/etc between different structures.

Reply:
I guess you wanted to say "Please add" instead of "Added"

Reply:
Well, partially added it myself :)